jre/java/java/util/stream/DoubleStream.java

```diff
@@ -147,19 +147,35 @@ public boolean tryAdvance(DoubleConsumer action) {
   }
 
   static DoubleStream iterate(double seed, DoubleUnaryOperator f) {
+    return iterate(seed, ignore -> true, f);
+  }
+
+  static DoubleStream iterate(double seed, DoublePredicate hasNext, DoubleUnaryOperator f) {
     Spliterator.OfDouble spliterator =
         new Spliterators.AbstractDoubleSpliterator(
             Long.MAX_VALUE, Spliterator.IMMUTABLE | Spliterator.ORDERED) {
+          private boolean first = true;
           private double next = seed;
+          private boolean terminated = false;
 
           @Override
           public boolean tryAdvance(DoubleConsumer action) {
+            if (terminated) {
+              return false;
+            }
+            if (!first) {
+              next = f.applyAsDouble(next);
```

Comment:
Looks like there is behavior change. In the original logic we were calling apply even for the first item. We are no longer doing that. Looking at the test, that seems intentional...

Reply:
@gkdn I have added tests that check this case